File: errand/backend/app/brokers/mock.py
"""Mock brokers — make the whole flow runnable with zero external calls.

Real brokers (prava.py, senso.py verified; shopper + mail pending) drop in via
the registry without touching the orchestrator.
"""

from __future__ import annotations

import time

from app.contracts import (
    CartItem,
    CartResult,
    CheckoutState,
    CreateSessionInput,
    CreateSessionResult,
    InboxMessage,
    OrderConfirmation,
    OrderResult,
    PaymentCredential,
    PollCompleted,
    PollCredentialResult,
    PollPending,
    ProfileKind,
    PurchaseContext,
    Merchant,
    Citation,
    TxnStatus,
)


class MockContextBroker:
    async def get_context(self, profile: ProfileKind, intent: str) -> PurchaseContext:
        if profile == "business":
            return PurchaseContext(
                profile=profile,
                approved_merchants=[Merchant(name="Demo Pantry Co", url="https://demo-pantry.example.com")],
                budget_cents=20000,
                rules=["Prefer Blue Bottle/Clif/LaCroix", "No energy drinks", "Under $200"],
                citations=[Citation(source="Procurement Policy v3", snippet="$200 cap; approved vendors only")],
            )
        return PurchaseContext(
            profile=profile,
            approved_merchants=[Merchant(name="Demo Pantry Co", url="https://demo-pantry.example.com")],
            budget_cents=6000,
            rules=["Oat milk, dark roast, sparkling water", "Low sugar", "Under $60"],
            citations=[Citation(source="My Preferences", snippet="$60 weekly; oat milk, dark roast")],
        )


class MockShopperBroker:
    async def build_cart(self, merchant_url, intent, context) -> CartResult:
        items = [
            CartItem(name="Blue Bottle Coffee 12oz", qty=2, price_cents=1800),
            CartItem(name="Clif Bars (12 pack)", qty=1, price_cents=1500),
            CartItem(name="LaCroix Sparkling Water (24)", qty=1, price_cents=1200),
        ]
        total = sum(i.qty * i.price_cents for i in items)
        while total > context.budget_cents and items:
            items.pop()
            total = sum(i.qty * i.price_cents for i in items)
        return CartResult(
            items=items,
            total_cents=total,
            checkout=CheckoutState(merchant_url=merchant_url, items=items, session_ref=f"mock-{int(time.time())}"),
        )

    async def complete_checkout(self, checkout, credential) -> OrderResult:
        oid = f"ORD-{int(time.time()) % 1_000_000}"
        return OrderResult(order_id=oid, confirmation_text=f"Order {oid} placed at {checkout.merchant_url}")


class MockPaymentBroker:
    def __init__(self) -> None:
        self._polls: dict[str, int] = {}

    async def create_session(self, data: CreateSessionInput) -> CreateSessionResult:
        sid = f"sess_mock_{int(time.time()*1000)}"
        return CreateSessionResult(session_id=sid, iframe_url=f"https://sandbox.prava.space/iframe?session={sid}")

    async def poll_credential(self, session_id: str) -> PollCredentialResult:
        n = self._polls.get(session_id, 0) + 1
        self._polls[session_id] = n
        if n < 2:
            return PollPending()
        return PollCompleted(
            credential=PaymentCredential(
                token="4111111111111111", dynamic_cvv="123",
                expiry_month="12", expiry_year="2029", txn_ref_id=f"txn_{session_id}",
            )
        )

    async def report_status(self, session_id: str, txn_ref_id: str, status: TxnStatus) -> None:
        return None


class MockMailBroker:
    _address = "errand-agent@demo.agentmail.to"

    async def ensure_inbox(self) -> str:
        return self._address

    async def wait_for_confirmation(self, merchant, since_iso, timeout_ms) -> OrderConfirmation:
        raw = InboxMessage(
            id=f"msg_{int(time.time())}", from_addr=f"orders@{merchant}",
            subject="Your order is confirmed",
            text="Thanks for your order! Order ORD-123456 received.",
            received_at=since_iso,
        )
        return OrderConfirmation(matched=True, order_id="ORD-123456", merchant=merchant, raw=raw)

    async def list_messages(self, limit: int = 10) -> list[InboxMessage]:
        return []

    async def reply(self, message_id: str, text: str) -> None:
        return None

File: errand/src/lib/brokers/mock.ts
import type {
  AuditEvent,
  AuditSink,
  ApprovalGate,
  CartResult,
  ContextBroker,
  CreateSessionInput,
  CreateSessionResult,
  InboxMessage,
  MailBroker,
  OrderConfirmation,
  OrderResult,
  PaymentBroker,
  PollCredentialResult,
  ProfileKind,
  PurchaseContext,
  ShopperBroker,
  TxnStatus,
} from "@/lib/contracts";
import { PROFILE_SEED } from "@/lib/profiles/seed";

/**
 * Mock implementations of every broker. These make the entire flow runnable
 * with zero API keys, and serve as the stubs parallel agents build the real
 * versions against. Each mock honours its interface exactly.
 */

// ── ContextBroker (mock Senso) ─────────────────────────────────────────────
export class MockContextBroker implements ContextBroker {
  async getContext(profile: ProfileKind, _intent: string): Promise<PurchaseContext> {
    // Real impl: query Senso KB with the intent, return cited rules.
    return structuredClone(PROFILE_SEED[profile]);
  }
}

// ── ShopperBroker (mock Stagehand/Browser Run) ─────────────────────────────
export class MockShopperBroker implements ShopperBroker {
  async buildCart(
    merchantUrl: string,
    _intent: string,
    context: PurchaseContext,
  ): Promise<CartResult> {
    // Real impl: drive the merchant with Stagehand act/extract/observe.
    const items = [
      { name: "Blue Bottle Coffee 12oz", qty: 2, priceCents: 1800 },
      { name: "Clif Bars (12 pack)", qty: 1, priceCents: 1500 },
      { name: "LaCroix Sparkling Water (24)", qty: 1, priceCents: 1200 },
    ];
    const totalCents = items.reduce((s, i) => s + i.qty * i.priceCents, 0);
    if (totalCents > context.budgetCents) {
      // Trim to respect budget — proves the guardrail in the mock too.
      items.pop();
    }
    const finalItems = items;
    const finalTotal = finalItems.reduce((s, i) => s + i.qty * i.priceCents, 0);
    return {
      items: finalItems,
      totalCents: finalTotal,
      checkout: {
        merchantUrl,
        items: finalItems,
        sessionRef: `mock-browser-${Date.now()}`,
      },
    };
  }

  async completeCheckout(
    checkout: { merchantUrl: string },
    _credential: unknown,
  ): Promise<OrderResult> {
    // Real impl: type token + dynamic CVV into the parked checkout page.
    const orderId = `ORD-${Math.floor(Math.random() * 1_000_000)}`;
    return {
      orderId,
      confirmationText: `Order ${orderId} placed at ${checkout.merchantUrl}`,
    };
  }
}

// ── PaymentBroker (mock Prava) ─────────────────────────────────────────────
export class MockPaymentBroker implements PaymentBroker {
  private polls = new Map<string, number>();

  async createSession(input: CreateSessionInput): Promise<CreateSessionResult> {
    const sessionId = `sess_mock_${Date.now()}`;
    return {
      sessionId,
      iframeUrl: `https://sandbox.prava.space/iframe?session=${sessionId}&amt=${input.totalCents}`,
    };
  }

  async pollCredential(sessionId: string): Promise<PollCredentialResult> {
    // Simulate the pending → completed transition Prava exhibits.
    const count = (this.polls.get(sessionId) ?? 0) + 1;
    this.polls.set(sessionId, count);
    if (count < 2) return { status: "pending" };
    return {
      status: "completed",
      credential: {
        token: "4111111111111111",
        dynamicCvv: "123",
        expiryMonth: "12",
        expiryYear: "2029",
        txnRefId: `txn_${sessionId}`,
      },
    };
  }

  async reportStatus(_sessionId: string, _txnRefId: string, _status: TxnStatus): Promise<void> {
    // Real impl: POST /v1/sessions/{id}/report-status
  }
}

// ── MailBroker (mock AgentMail) ────────────────────────────────────────────
export class MockMailBroker implements MailBroker {
  private address = "errand-agent@demo.agentmail.to";

  async ensureInbox(): Promise<{ address: string }> {
    return { address: this.address };
  }

  async waitForConfirmation(opts: {
    merchant: string;
    sinceIso: string;
    timeoutMs: number;
  }): Promise<OrderConfirmation> {
    // Real impl: poll inboxes.messages.list until the confirmation arrives.
    const raw: InboxMessage = {
      id: `msg_${Date.now()}`,
      from: `orders@${opts.merchant.replace(/^https?:\/\//, "")}`,
      subject: "Your order is confirmed",
      text: "Thanks for your order! Order ORD-123456 has been received.",
      receivedAt: new Date().toISOString(),
    };
    const orderId = raw.text.match(/ORD-\d+/)?.[0];
    return { matched: true, orderId, merchant: opts.merchant, raw };
  }

  async listMessages(_limit = 10): Promise<InboxMessage[]> {
    return [];
  }

  async reply(_messageId: string, _text: string): Promise<void> {
    // Stretch.
  }
}

// ── ApprovalGate (auto-approve for tests/CLI) ──────────────────────────────
export class AutoApproveGate implements ApprovalGate {
  async requestApproval(): Promise<{ approved: boolean }> {
    return { approved: true };
  }
}

// ── AuditSink (in-memory) ──────────────────────────────────────────────────
export class MemoryAuditSink implements AuditSink {
  private events: AuditEvent[] = [];
  record(event: AuditEvent): void {
    this.events.push(event);
  }
  all(): AuditEvent[] {
    return [...this.events];
  }
}
